File: java-call-clj/src/main/java/capclug/demo/callclj/Eight.java
/*
 * $Id$
 * Copyright (C) 2010 ARIN
 */
package capclug.demo.callclj;

import capclug.demo.cljfromjava.eight;

/**
 * @version $Rev$
 */
public class Eight
{
    public static void main( String[] args )
    {
        eight eight = new eight();
        eight.foo( 42d ); // takes any object!
        eight.bar( 43d ); // runtime exception! tries to cast Double to String
    }
}

File: compile-clj-to-java/src/main/clojure/capclug/demo/cljfromjava/eight.clj
(ns capclug.demo.cljfromjava.eight
  (:gen-class
    :methods [[foo [Object] void]
              [bar [Object] void]
              [baz [Object] void]]))

(defn -foo [this #^String baz]
  (println "will I work? " baz))

; tries to cast Double to String
(defn -bar [this #^String baz]
  (println "will I work? " (.toUpperCase baz)))
